compiler/qsc_passes/src/capabilitiesck.rs

```diff
@@ -0,0 +1,555 @@
+        self.visit_expr(condition_expr_id);
+        self.visit_block(body_block_id);
+        let post_spans_with_missing_features_count = self.missing_features_map.len();
+
+        // If no errors were added because of the individual elements, check this expression as a whole.
+        let errors_delta =
+            post_spans_with_missing_features_count - pre_spans_with_missing_features_count;
+        if errors_delta == 0 {
+            self.check_expr(expr_id);
+        }
+    }
+
+    fn check_spec_decl(
+        &mut self,
+        functor_set_value: FunctorSetValue,
+        spec_decl: Option<&'a SpecDecl>,
+    ) {
+        let current_callable_id = self.get_current_callable();
+        let ItemComputeProperties::Callable(callable_compute_properties) =
+            self.compute_properties.get_item(current_callable_id)
+        else {
+            panic!("expected callable variant of item compute properties");
+        };
+
+        let spec_compute_properties = match functor_set_value {
+            FunctorSetValue::Empty => &callable_compute_properties.body,
+            FunctorSetValue::Adj => callable_compute_properties
+                .adj
+                .as_ref()
+                .expect("adj specialization is none"),
+            FunctorSetValue::Ctl => callable_compute_properties
+                .ctl
+                .as_ref()
+                .expect("ctl specialization is none"),
+            FunctorSetValue::CtlAdj => callable_compute_properties
+                .ctl_adj
+                .as_ref()
+                .expect("ctl_adj specialization is none"),
+        };
+
+        if let ComputeKind::Quantum(quantum_properties) = spec_compute_properties.inherent {
+            let missing_features = get_missing_runtime_features(
+                quantum_properties.runtime_features,
+                self.target_capabilities,
+            );
+            let missing_spec_level_runtime_features =
+                get_spec_level_runtime_features(missing_features);
+
+            // If there are any missing specialization-level runtime features, runtime features that affect the whole
+            // specialization, just generate errors for the missing specialization-level runtime features and do not
+            // generate statement-level or expression-level errors for these specializations.
+            if !missing_spec_level_runtime_features.is_empty() {
+                let current_callable = self
+                    .package
+                    .get_global(current_callable_id)
+                    .expect("callable not present in package");
+                let Global::Callable(callable_decl) = current_callable else {
+                    panic!("");
+                };
+
+                self.missing_features_map
+                    .entry(callable_decl.name.span)
+                    .and_modify(|f| *f |= missing_spec_level_runtime_features)
+                    .or_insert(missing_spec_level_runtime_features);
```

Comment:
You don't actually need to use itertools for this, and can instead use built-in stdlib stuff as long as `RuntimeFeatureFlags` gets support for `Default`:

```suggestion
                *(self
                    .missing_features_map
                    .entry(callable_decl.name.span)
                    .or_default()) |= missing_spec_level_runtime_features;
```

Reply:
This part of the code does not need itertools. The place where I am using it is [here](https://github.com/microsoft/qsharp/blob/644bacd3436fc78cbd32ed37f6d1904eae91dc1f/compiler/qsc_passes/src/capabilitiesck.rs#L437), to being able to always produce the errors in an expected order (otherwise tests are flaky because of the ordering).

Reply:
Ah, I commented on that line then. I think it would be better to avoid itertools, as it makes code look nicer but introduces unexpected costs.

Reply:
Removed itertools dependency using your suggestion :)